grouper/fe/handlers.py

```diff
@@ -858,6 +888,172 @@ def get(self, group_id=None, name=None):
+        if not group:
+            return self.notfound()
+
+        args_by_perm = get_grantable_permissions(self.session)
+        dropdown_form, text_form = self._get_forms(args_by_perm, get_request_data=False)
+
+        self.render("group-permission-request.html", dropdown_form=dropdown_form,
+                text_form=text_form, group=group, args_by_perm_json=json.dumps(args_by_perm))
+
+    def post(self, group_id=None, name=None):
+        group = Group.get(self.session, group_id, name)
+        if not group:
+            return self.notfound()
+
+        # only owner of group can request permissions for that group
+        role_index = self.current_user.my_role_index(group.my_members())
+        if role_index not in OWNER_ROLE_INDICES:
+            return self.forbidden()
+
+        # check inputs
+        args_by_perm = get_grantable_permissions(self.session)
+        dropdown_form, text_form = self._get_forms(args_by_perm, get_request_data=True)
+
+        argument_type = self.request.arguments.get("argument_type")
+        if argument_type and argument_type[0] == "text":
+            form = text_form
+        elif argument_type and argument_type[0] == "dropdown":
+            form = dropdown_form
+            form.argument.choices = [(a, a) for a in args_by_perm[form.permission_name.data]]
+        else:
+            # someone messing with the form
+            return self.forbidden()
+
+        if not form.validate():
+            return self.render(
+                    "group-permission-request.html", dropdown_form=dropdown_form,
+                    text_form=text_form, group=group, args_by_perm_json=json.dumps(args_by_perm),
+                    alerts=self.get_form_alerts(form.errors),
+                    )
+
+        permission = Permission.get(self.session, form.permission_name.data)
+        assert permission, "our prefilled permission should exist or we have problems"
+
+        # save off request
+        try:
+            permissions.create_request(self.session, self.current_user, group,
+                    permission, form.argument.data, form.reason.data)
+        except permissions.RequestAlreadyGranted:
+            alerts = [Alert("danger", "this group already has this permission + argument")]
+        except permissions.RequestAlreadyExists:
+            alerts = [Alert("danger", "request for permission + argument already exists")]
+        except permissions.NoOwnersAvailable:
+            alerts = [Alert("danger", "no owners available for requested permission + argument")]
+        else:
+            alerts = None
+
+        if alerts:
+            return self.render(
+                    "group-permission-request.html", dropdown_form=dropdown_form,
+                    text_form=text_form, group=group, args_by_perm_json=json.dumps(args_by_perm),
+                    alerts=alerts,
+                    )
+        else:
+            return self.redirect("/groups/{}".format(group.name))
```

Comment:
yes, but it's probably not great right now. we show requests at the bottom as pending in the permission panel.

i have a todo to make this better (threading messages through redirects).